File: algorithms/123.Best_Time_to_Buy_and_Sell_Stock_III/Best_Time_to_Buy_and_Sell_Stock_III.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int maxProfit(vector<int> &prices) {
        int n = prices.size();
        if (n == 0)
            return 0;
        vector<int> left(n);
        int buy_price = prices[0], profit = 0;

        // pre deal the left
        for (int i = 1; i < n; ++i) {
            buy_price = min(buy_price, prices[i]);
            profit = max(profit, prices[i] - buy_price);
            left[i] = profit;
        }

        int ret = left[n - 1];
        buy_price = prices[n - 1];
        profit = 0;
        // deal the right and get result
        for (int i = n - 1; i >= 1; --i) {
            buy_price = max(buy_price, prices[i]);
            profit = max(profit, buy_price - prices[i]);
            ret = max(ret, profit + left[i - 1]);
        }
        return ret;
    }
};

class Solution {
public:
    /**
     * @param prices: Given an integer array
     * @return: Maximum profit
     */
    int maxProfit(vector<int> &prices) {
        if (prices.size() == 0) {
            return 0;
        }
        
        vector<int> profit(prices.size());
        int buy = 0;
        buy = prices[0];
        profit[0] = 0;
        for (int i = 1; i < prices.size(); i++) {
            profit[i] = max(profit[i - 1], prices[i] - buy);
            buy = min(buy, prices[i]);
        }
        
        int sell = prices[prices.size() - 1];
        int best = 0;
        for (int i = prices.size() - 2; i >= 0; i--) {
            best = max(best, sell - prices[i] + profit[i]);
            sell = max(sell, prices[i]);
        }
        
        return best;
    }
};

int main() {
    int n;
    cin >> n;
    vector<int> arr(n);
    for (auto & i : arr)
        cin >> i;
    Solution s;
    cout << s.maxProfit(arr) << endl;
    return 0;
}

File: algorithms/123.Best_Time_to_Buy_and_Sell_Stock_III/Best_Time_to_Buy_and_Sell_Stock_III.py
class Solution:
    def maxProfit(self, prices):
        buy1, buy2 = -(1<<31), -(1<<31)
        sell1, sell2 = 0, 0

        for i in prices:
            sell2 = max(sell2, buy2 + i)    # max money after sell 2
            buy2 = max(buy2, sell1 - i)     # max money after buy 2
            sell1 = max(sell1, buy1 + i)    # max money after sell 1
            buy1 = max(buy1, -i)            # max money after buy 1

        return sell2


# 1.dp
# 交易两次
class Solution2:
    def maxProfit(self, prices):
        if not prices:
            return 0
        n = len(prices)
        # 初始化状态
        dp = [[[0]*2 for _ in range(3)] for _ in range(n)]
        for k in range(3):
            dp[0][k][1] = -prices[0]
        # 从 i=1 处开始迭代
        for i in range(1, n):
            for k in range(1, 3):
                dp[i][k][0] = max(dp[i-1][k][0], dp[i-1][k][1]+prices[i])
                dp[i][k][1] = max(dp[i-1][k][1], dp[i-1][k-1][0]-prices[i])
        return dp[n-1][2][0]


class Solution2:
    def maxProfit(self, prices):
        if len(prices) == 0:
            return 0
        dp = [[0] * (len(prices)) for _ in range(3)]

        for i in range(1,3):
            for j in range(1, len(prices)):
                profit = 0
                for m in range(j):
                    profit = max(profit, dp[i-1][m]+prices[j]-prices[m])
                dp[i][j] = max(dp[i][j-1], profit)

        return dp[-1][-1]


if __name__ == "__main__":
    result = Solution2().maxProfit([3, 2, 1, 4, 2, 5, 6])
    print(result)
